applications/MeshMovingApplication/custom_utilities/explicit_mesh_moving_utilities.cpp

```diff
@@ -0,0 +1,254 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//
+//  Main authors:    Ruben Zorrilla
+//
+//					 Kratos default license: kratos/license.txt
+//
+
+// System includes
+
+// External includes
+
+// Application includes
+#include "move_mesh_utilities.h"
+#include "explicit_mesh_moving_utilities.h"
+
+// Project includes
+#include "spatial_containers/spatial_containers.h"
+#include "utilities/binbased_fast_point_locator.h"
+#include "utilities/spatial_containers_configure.h"
+
+namespace Kratos
+{
+    /* Public functions *******************************************************/
+
+    ExplicitMeshMovingUtilities::ExplicitMeshMovingUtilities(
+        ModelPart &rModelPart,
+        ModelPart &rStructureModelPart,
+        const double SearchRadius) : 
+        mSearchRadius(SearchRadius),
+        mrVirtualModelPart(rModelPart),
+        mrStructureModelPart(rStructureModelPart){
+
+        KRATOS_WARNING_IF("ExplicitMeshMovingUtilities",mrStructureModelPart.GetBufferSize() == 1) << 
+            "Structure model part buffer size is 1. Setting buffer size to 2." << std::endl;
+        mrStructureModelPart.SetBufferSize(2);
+    }
+
+    void ExplicitMeshMovingUtilities::FillVirtualModelPart(ModelPart& rOriginModelPart){
+
+        // Check that the origin model part has nodes and elements to be copied
+        KRATOS_ERROR_IF(rOriginModelPart.NumberOfNodes() == 0) << "Origin model part has no nodes.";
+        KRATOS_ERROR_IF(rOriginModelPart.NumberOfElements() == 0) << "Origin model part has no elements.";
+
+        // Set the buffer size in the virtual model part
+        mrVirtualModelPart.SetBufferSize(rOriginModelPart.GetBufferSize());
+
+        // Copy the origin model part nodes and fixity
+        auto &r_nodes_array = rOriginModelPart.NodesArray(); 
+        for(auto it_node : r_nodes_array){
+            // Create a copy of the origin model part node
+            auto p_node = mrVirtualModelPart.CreateNewNode(it_node->Id(),*it_node, 0);
```

Comment:
I think this is expensive
Do you really have to copy them or could you use the existing one

Reply:
why do you get the "NodesArray" however?

i would do

      for(auto& it_node : rOriginModelPart.Nodes())
                  .... it_node.Id()

Reply:
The auto loop cannot be used with the Nodes() container. Maybe we can do it in the common way...